src/Microsoft.Extensions.Options/OptionsSnapshot.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Implementation of IOptionsSnapshot.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsSnapshot<TOptions> : IOptionsSnapshot<TOptions> where TOptions : class, new()
+    {
+        private readonly IOptionsCache<TOptions> _cache;
+        private readonly IOptionsFactory<TOptions> _factory;
+        private readonly ConcurrentDictionary<string, TOptions> _snapshots = new ConcurrentDictionary<string, TOptions>();
+
+        /// <summary>
+        /// Initializes a new instance with the specified options configurations.
+        /// </summary>
+        /// <param name="cache">The cache to use.</param>
+        /// <param name="factory">The factory to use to create options.</param>
+        /// <param name="changeSources">The change token sources used to detect options changes.</param>
+        public OptionsSnapshot(IOptionsCache<TOptions> cache, IOptionsFactory<TOptions> factory, IEnumerable<IOptionsChangeTokenSource<TOptions>> changeSources)
```

Comment:
E.g. if the `IEnumerable<IConfigureOptions<TOptions>>` argument in the constructor of the factory was also resolved and instantiated every time, then I believe the cost would be significant.